### DISCUSSION TITLE
CommonJS support

### DISCUSSION THREAD

**Discussion by ericvergnaud** [antlr/antlr4#4215] [Ideas]
Currently antlr only supports ESM. 
This is an issue for developers tied to now legacy CommonJS
The problem to solve is more complex for antlr than for most libraries, because antlr generates code for ESM only, not for CJS. If a solution can be identified that is lightweight enough and non-destructive we will consider it. 
This discussion is about that (and only that).
Please don't submit PRs until this discussion reaches some level of approval from antlr org (they will be redirected here).

**Comment by ericvergnaud** [comment]
The ideal solution is one that wraps the official package and exposes it as a CJS module without requiring any changes in the official package.

The rationale is 2-fold:
 - CJS will ultimately disappear, and we don't want to pollute the build chain and code base with it
 - I'm not able to invest time in supporting CJS and the 3 dozen packaging tools out there

**Comment by ericvergnaud** [comment]
The solution could be as lightweight as providing a sample project for exposing antlr + generated parsers as a CJS package. We can then include that sample in the docs.

**Comment by ericvergnaud** [comment]
see #4217

**Reply by ericvergnaud** [reply]
see #4193

**Reply by bugwelle** [reply]
#4217 works for us. I updated from Antlr 4.9.2 (last version to use CommonJs if I remember correctly) to v4.13.0. For reference:

```sh
# Antlr 4.13.0
du -sh  node_modules/antlr4         
# 6.7M	node_modules/antlr4

# Antlr 4.9.2
du -sh  node_modules/antlr4
# 524K	node_modules/antlr4
```

For v4.13.0 >5MiB are due to `dist/` with all its variations and map files.

**Reply by ericvergnaud** [reply]
@bugwelle Thanks for confirming. I'm closing this discussion accordingly.

**Comment by ericvergnaud** [comment]
see #4199
